teps/0052-nested-triggers.md

```diff
@@ -0,0 +1,202 @@
+
+### Goals
+
+* Allow some shared logic to be processed before Triggers are called.
+* Provide a way to exclude some `Triggers` from processing by the `EventListener` HTTP handler
+* Pass the same data from an interceptor chain processing down to multiple `Triggers`. This is in line with
+  recent discussions around only allowing modifications to extensions, rather than the current format
+  where webhook extensions can modify the body and headers of the incoming event.
+* Simplify the maintenance around branching logic in triggering. For example, if the logic for two
+  Triggers is similar, but dependent on a single parameter, those two pipelines can have the same
+  filter set defined in the upstream Trigger. Then, the two pipelines filters only need to specify
+  the differentiating logic.
+
+### Non-Goals
+
+* While we posit that performance improvements are possible by limiting the number of invoked triggers,
+  significant performance gains is not intended to be a goal of this proposal.
+
+### Use Cases (optional)
+
+* Add common metadata to every github event based on external contextual information
+* Simplify the writing of new Triggers by allowing operators to provide reusable interceptor chains
+* Filter events that don't pass validation in a single Trigger to reduce noise and ease debugging
+* Generate a single PVC name for a set of pipelines to use based on event context and pass PVC 
+  into each pipelinerun as a workspace
+
+## Requirements
+
+* Users should be able to reuse a single chain of interceptors across multiple triggers
+* Users should be able to invoke multiple triggers after interceptor processing completes
+* Users should be able to exclude a set of triggers from event listener HTTP request processing
+
+## Proposal
+
+This document has been opened to begin a discussion around the feature. There is a draft implementation
+in [triggers](https://github.com/tektoncd/triggers/pull/946). Our team from Optum is also willing
+to come demo the capabilities of this feature in a broader working group meeting, if desired.
+
+### Implementation Notes
+
+The current implementation provides the following set of API changes:
+
+* New `triggers` filed would be optional within `TriggerSpec` and accept any unicode character.
+* New `triggers` field would be optional within `EventListenerSpec` and accept any unicode character.
+* Update validation on these two objects to allow for `TriggerTemplate` to be unset
+
+One of the key challenges here is that if the logic for a downstream trigger is simplified, then this
+downstream trigger must be excluded from processing in the default HTTP endpoint. One of the goals
+outlined above is that Trigger authors are enabled to focus on the logic specific to their trigger.
+If an interceptor chain is removed from the processing of an authored trigger then it cannot be
+exposed for direct requests otherwise requests could bypass the desired filtering.
```

Comment:
Haven't fully thought through this, but an idea I'm thinking about is modeling this as an authorization problem and using that as a mechanism for determining whether an entity can expose / invoke a given trigger.

I haven't fully thought through the identities at play here (e.g. is it only the EL, or do triggers have their own identity?), but I think this might help model some of the more complex cases of multiple ELs referencing the same trigger.

Reply:
@wlynch I do like the idea of having the event listener specify a trigger selector. right now, the only option is a namespace selector and then it queries all triggers. given the current implementation around querying triggers, this would be a very simple extension that would enable event listener operators to specify the triggers that should be processed.

Reply:
This is a very interesting discussion, we can probably merge this one as it is now, and continue the discussion on this on the next PR?